src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs

```diff
@@ -181,6 +189,34 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
             return locals?.ToImmutableAndFree() ?? ImmutableArray<LocalSymbol>.Empty;
         }
 
+        // When a VariableDeclaration is used in a deconstruction, there are two cases:
+        // - deconstruction is set, type may be set (for "var"), and no declarators. For instance, `var (x, ...)` or `(int x, ...)`.
+        // - deconstruction is null, type may be set, and there is one declarator holding the identifier. For instance, `int x` or `x`.
+        private void CollectLocalsFromDeconstruction(VariableDeclarationSyntax declaration, TypeSyntax varSyntax, ArrayBuilder<LocalSymbol> locals)
+        {
+            if (declaration.Deconstruction != null)
+            {
+                var nestedVarSyntax = varSyntax ?? declaration.Type;
+                foreach (var variable in declaration.Deconstruction.Variables)
+                {
+                    CollectLocalsFromDeconstruction(variable, nestedVarSyntax, locals);
+                }
+            }
+            else
+            {
+                Debug.Assert(declaration.Variables.Count == 1);
+                var declarator = declaration.Variables[0];
+
+                SourceLocalSymbol localSymbol = SourceLocalSymbol.MakeDeconstructionLocal(
+                                                            this.ContainingMemberOrLambda,
+                                                            this,
+                                                            declaration.Type,
+                                                            declarator.Identifier,
+                                                            isVar: varSyntax != null);
```

Comment:
I think the closest type syntax should win. If declaration.Type is not null, it should be used, otherwise varSyntax should be used. But we should still check if varSyntax is actually a var syntax.

---
In reply to: [68807675](https://github.com/dotnet/roslyn/pull/12224#discussion_r68807675) [](ancestors = 68807675)